src/broadway/QueuePublishingEventBus.php

```diff
@@ -0,0 +1,72 @@
+<?php
+namespace Burrow\Broadway;
+
+use Broadway\Domain\DomainEventStreamInterface;
+use Broadway\EventHandling\SimpleEventBus;
+use Burrow\QueuePublisher;
+
+class QueuePublishingEventBus extends SimpleEventBus
+{
+    /**
+     * @var DomainEventStreamInterface[]
+     */
+    private $queue;
+
+    /**
+     * @var bool
+     */
+    private $isPublishing;
+
+    /**
+     * @var DomainEventStreamSerializer
+     */
+    private $serializer;
+
+    /**
+     * @var QueuePublisher
+     */
+    private $queuePublisher;
+
+    /**
+     * Constructor
+     *
+     * @param DomainEventStreamSerializer $serializer
+     * @param QueuePublisher              $queuePublisher
+     */
+    public function __construct(DomainEventStreamSerializer $serializer, QueuePublisher $queuePublisher)
+    {
+        $this->queue = [];
+        $this->isPublishing = false;
+        $this->serializer = $serializer;
+        $this->queuePublisher = $queuePublisher;
+    }
+
+    /**
+     * Publishes the events from the domain event stream to the listeners.
+     *
+     * @param DomainEventStreamInterface $domainMessages
+     * @throws \Exception
+     */
+    public function publish(DomainEventStreamInterface $domainMessages)
+    {
+        $this->queue[] = $domainMessages;
+
+        if (! $this->isPublishing) {
```

Comment:
It's here to ensure events are dispatched in the order they've been published.

If you misuse the bus and publish a new event inside a listener, without it, the bus will dispatch the new event before the current one has finished dispatching it, resulting in a very unpredictable behavior.

Adding it to a queue, and knowing if an event is being dispatched solves this issue.

So, I've kept this behavior which exists in the parent class. I've also found it in Tactician for instance.

But yeah, first time I saw this, I was also puzzled.